### PR TITLE
Add MPU information to panic!() print

### LINKED ISSUE #925: Tracking: MPU Configuration Print on Panic
### Goal

The kernel panic print should include something like:

```
MPU Configuration:
 0: 0x00030000-0x00030200 Sub ✘'s 0x00 S ✘ X ✔ B ✘ C ✘ TEX 0 AP 6
 1: 0x20004000-0x20006000 Sub ✘'s 0x00 S ✘ X ✔ B ✘ C ✘ TEX 0 AP 3
 2: Disabled.
 3: Disabled.
 4: Disabled.
 5: Disabled.
 6: Disabled.
 7: Disabled.
```

### Progress

Branch: https://github.com/tock/tock/tree/mpu-on-panic

### Todo

- [ ] Cache MPU region configurations.
- [ ] Print the cache on panic.

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request changes the `MpuConfig` type to require that it implement the `fmt::Display` trait so that the process.rs panic printout can display the MPU information.

This fixes #925.

### Testing Strategy

This pull request was tested by running `crash_dummy` on Hail and verifying that the MPU display looks ok.

### TODO or Help Wanted

1. The output looks like:

    ```
     Cortex-M MPU
      Region 0: base:  20004000, length: 8192 bytes; ReadWrite (0x3)
      Region 1: base:     30000, length: 1024 bytes; ReadOnly (0x6)
      Region 2: Unused
      Region 3: Unused
      Region 4: Unused
      Region 5: Unused
      Region 6: Unused
      Region 7: Unused
    ```

    I'm not sure how to get the `0x` to display on the base address. The format string looks like `Region {}: base:{:>width$x}` and sticking a `#` in there to make it print the `0x` gives me errors.

2. The rust compiler generates an error when linking for risc-v. https://github.com/rust-lang/rust/issues/65774 I have no idea what to do about that.

### Documentation Updated

I added a new copy of the panic!() output to the docs to keep that up to date.

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by alevy** [comment]
> I'm not sure how to get the 0x

A hack might be to just explicity add "0x" to the format string

**Comment by bradjc** [comment]
I keep trying on new versions of Rust and am still getting the same ICE.

**Comment by bradjc** [comment]
Ok! Ready for review.

File: kernel/src/platform/mpu.rs [hunk 1]
```diff
@@ -49,6 +49,20 @@ impl Region {
     }
 }
 
+/// Null type for the default type of the `MpuConfig` type in an implementation
+/// of the `MPU` trait. We need this to workaround a bug in the Rust compiler.
+///
+/// If https://github.com/rust-lang/rust/issues/65774 is resolved then we should
```
Comment (pnkfelix) [inline kernel/src/platform/mpu.rs]: My assumption is that the resolution of rust-lang/rust#65774 is not going to include making `()` implement `Display`.

That is, if that bug is fixed and you revert this commit, then what you will get is a proper compiler error at the point where you are doing:

```rust
type MpuConfig: Default + Display = ();
```

My suggestion is that you just keep this `MpuConfigDefault` type. (That, or use something other than `fmt::Display` as your bound on the `MpuConfig` associated type.)

**Pull request merged**